### DISCUSSION TITLE
Make static factory methods possible.

### DISCUSSION THREAD

**Discussion by procodix** [dotnet/csharplang#8243] [Language Ideas]
```c#
public class A {
	public static A Factory() {
		return new A();
	}
}

public class B : A {}
```

B::Factory() returns an A not a B, which is a violation of inheritance. 

There should be a new keyword like "This". During compilation all static methods with this keyword have to be copied for all subtypes and "This" has to be replaced with the Copy's type. That saves the burden of manually copying the method into every class.

```c#
public class A {
	public static This Factory() {
		return new This();
	}
}

public class B : A {
       //compiler generated copy of A.Factory() :
	public static B Factory() {
		return new B();
	}
}
```

**Comment by bernd5** [comment]
You may be able to accomplish this by a source generator.

**Reply by procodix** [reply]
How would that work?

**Reply by bernd5** [reply]
You would need to make your classes partial.
Then you need something which controls where and what code should be generated. Often this is done with attributes (but you could look for a pattern, too - e.g. static method named "Factory" returning the containing type.

You could write e.g.:

```
public class A {
        [DynFactory] // this may controls the source generator (could be located in an another assembly)
	public static A Factory() {
		return new A();
	}
}

public partial class B : A 
{      
}
```

Then the generator could create something like this:
```
partial class B
{
	public static B Factory() {
		return new B();
	}
}
```
(containing type replaced with the new containing type)

**Reply by procodix** [reply]
Thats sounds way too complicated compared to compiler support. A keyword is cleaner and preferred over attributes. Plus: no refactoring, IDE support, ...

**Comment by HaloFour** [comment]
See: #5413

**Reply by procodix** [reply]
That is only for interfaces and it looks cluttered. Why not a clean syntax?

**Reply by HaloFour** [reply]
It has nothing to do with interfaces, it has to do with generics.  It's a much more applicable form of this feature.

As for the syntax you suggested, it implies that static methods are virtual and inherited, neither of which are true.  When you call `B::Factory`, that is purely compiler candy for `A::Factory`.  There is no `B::Factory`, nor is there a mechanism for that method to know that you had intended to create a `B`, aside from generics.  Beyond that, there's no constraint on the constructor on the derived type, so `new This()` doesn't make sense since `This` could be declared without a parameterless constructor.

**Reply by procodix** [reply]
The way it works now is the worst possible compromise. It looks like inheritance but it does the opposite. 

1. What ist the use for the "compiler candy", I mean who would fool himself by even writing B::Factory, when it doesn't exist at all, and in fact it is actually A::Factory? Everybody writes A::Factory then.
2. My suggestion is, that there *should be* a compiler generated B::Factory which can be overriden individually. That would at least make some sense.

Don't overthink this into inheritance, constraints. When A has a constraint, B has it as well. Thats not a problem at all.

The proposal is very simple. Copy the method and replace the return types. That is what happens when you copy and paste the method manually.

**Reply by yaakov-h** [reply]
"copying the method" is impossible due to ref assemblies.

**Reply by procodix** [reply]
Why? Its a forwarw-only-feature for methods containing the new keyword. Even if you'd write a new app with B::Factory which overrides an ancient assembly's A::Factory, the compiler could generate code to call A::Factory and type cast the result to a B.

**Reply by HaloFour** [reply]
> and type cast the result to a B.

Which would fail because the instance is not a `B`.

**Reply by procodix** [reply]
but covariant to B

**Reply by HaloFour** [reply]
Which doesn't change that the instance is not a `B` and cannot be cast to `B`.  This is why the generic aspect of the linked proposal is necessary.

**Comment by FaustVX** [comment]
> The proposal is very simple. Copy the method and replace the return types. That is what happens when you copy and paste the method manually.

What happens if your derived `class` has not he same parameters as the base `class`:
```C#
class Animal(int numOfLegs) {
  public static Animal Factory(int numOfLegs) => new Animal(numOfLegs);
}

class Cat(Color furColor) : Animal(4);
```

**Reply by procodix** [reply]
First: You are thinking it completely wrong. Instead of hardcoded class names, you would use *This* keyword
Second: Your example indeed enforces a constraint to have a specific constructor. Thats totally not a problem to detect. However it would be much easier to use a parameterlesss constructor constraint new(), (compare Swift documentation for requirements of a single parameterless constructor, and constructor chaining), and assign the property manually. The constructor property promotion is a reduced, bad replacement for real universal static methods as factories.

```c#
class Animal(int numOfLegs) {
  // this gets copied to Cat class. This gets replaces by the current class
  public static *This* Factory(int numOfLegs) => new *This*(numOfLegs);
}

class Cat(Color furColor) : Animal(4);

// this is, how Swift an PHP do it:
class Cat {
  // copy from anchestor
  public static *This* Factory(int numOfLegs) => new *This*(numOfLegs);
  // call inherited static Factory method. 
  public static *This* Factory(int numOfLegs = 4, int color = "white") => base/This.Factory(numOfLegs).ColorSet(color)
}
```
